text/0000-io-net-2.1.md

```diff
@@ -0,0 +1,527 @@
+        pub interface: u32,
+    }
+}
+
+mod os::unix::net {
+    pub enum TcpKeepalive {}    // TCP_KEEP{ALIVE,IDLE}, Value = Duration
+    // others as necessary
+}
+
+impl {Socket,TcpListener,TcpStream,UdpSocket} {
+    pub fn set<T: Sockopt>(&self, val: T::Value) -> io::Result<()> {
+        T::set(self, val)
+    }
+
+    pub fn get<T: Sockopt>(&self) -> io::Result<T::Value> {
+        T::get(self)
+    }
+}
+
+// example usage
+use std::net::opt::{Keepalive, TcpKeepalive};
+
+socket.set::<Keepalive>(true);
+socket.set::<TcpKeepalive>(duration);
+```
+
+With this API there's a strict transcribing of the conventions proposed above
+with the `Sockopt` trait, and the major upside of this approach is that
+libraries can define their own socket options (via a new type) which can use the
+same syntax as all the standard socket options for being get/set. There are,
+however, a number of downsides to this approach:
+
+* Setting an option requires at least one import.
+* Socket options are now grouped together, but discoverability isn't necessarily
+  improved as there are still lots of implementors that need to be waded
+  through.
+* It's not possible to statically prevent TCP options from being set on a UDP
+  socket (which will likely always generate an error at runtime).
+* Lots of new types are introduced just for adding names, which is somewhat
+  unconventional in the standard library right now.
+
+### `TcpListener::bind_to_port`
+
+Requested, in [the std::net expansion issue][issue], it's noted that a common
+desire is to simply bind a listener to a port not worrying about the address.
+This RFC does not at this time add this sort of convenience method.
+
+### Enhancing `lookup_host` and `lookup_addr`
+
+These two bindings of `getaddrinfo` and `getnameinfo` are currently unstable yet
+are quite useful from time to time. This RFC does not propose stabilizing these
+APIs or reconsidering them at this time. The possible known drawbacks of the API
+are:
+
+* The `getaddrinfo` and `getnameinfo` function calls have a number of arguments
+  which are not exposed at this layer of the API (e.g. flags, `serv`, `host`,
+  `hints`, etc). Additionally, the return value for `getaddrinfo` is not fully
+  exposed.
+
+* The return value of these APIs can be dubious by turning up duplicates.
+
+* There is a desire to possibly filter the return value based on certain
+  criteria.
+
+Overall these functions may want to be tweaked slightly to have different
+arguments and/or return values after the underlying implementation has been
+investigated more. At this time this RFC does not strive to handle these APIs.
+
+### Asynchronous I/O
+
+Networking primitives are one of the primary users of asynchronous I/O
+utilities, and there are clear cross-platform abstractions for building this
+sort of functionality such as `select` and setting a socket to nonblocking mode.
+
+This RFC does not, however, attempt to set forth a vision for nonblocking
```

Comment:
Pity that. I'd have thought this would be the first extension point of any `std::net` overhaul.